pkg/kubelet/cm/deviceplugin/endpoint_handler.go

```diff
@@ -0,0 +1,224 @@
+	NewEndpoint(resourceName, socketPath string) error
+
+	Devices() map[string][]pluginapi.Device
+	Endpoint(resourceName string) (endpoint, bool)
+
+	Stop() error
+
+	Store() endpointStore
+	SetStore(endpointStore)
+
+	Callback() monitorCallback
+	SetCallback(f monitorCallback)
+}
+
+type endpointStore interface {
+	Endpoint(resourceName string) (endpoint, bool)
+	SynchronizedEndpoint(resourceName string) (*synchronizedEndpoint, bool)
+
+	SwapEndpoint(e endpoint) (endpoint, bool)
+	DeleteEndpoint(resourceName string) error
+
+	Range(func(string, *synchronizedEndpoint))
+}
+
+type synchronizedEndpoint struct {
+	e    endpoint
+	done chan string
+}
+
+type endpointStoreImpl struct {
+	sync.Mutex
+	endpoints map[string]*synchronizedEndpoint
+}
+
+type endpointHandlerImpl struct {
+	store    endpointStore
+	callback monitorCallback
+}
+
+func newEndpointHandlerImpl(f monitorCallback) *endpointHandlerImpl {
+	return &endpointHandlerImpl{
+		store:    newEndpointStoreImpl(),
+		callback: f,
+	}
+}
+
+func (h *endpointHandlerImpl) NewEndpoint(resourceName, socketPath string) error {
+	var devStore deviceStore = newDeviceStoreImpl(h.callback)
+	if old, ok := h.store.Endpoint(resourceName); ok && old != nil {
+		devStore = old.Store()
+	}
+
+	e, err := newEndpoint(socketPath, resourceName, devStore)
+	if err != nil {
+		return err
+	}
+
+	old, ok := h.store.SwapEndpoint(e)
+	go h.trackEnpoint(e)
+
+	if ok && old != nil {
+		old.SetStore(newAlwaysEmptyDeviceStore())
+
+		if err = old.Stop(); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (h *endpointHandlerImpl) trackEnpoint(e endpoint) {
+	e.Run()
+
+	if curr, ok := h.store.SynchronizedEndpoint(e.ResourceName()); ok && curr.e == e {
+		h.store.DeleteEndpoint(e.ResourceName())
+
+		if curr.done != nil {
+			curr.done <- e.ResourceName()
+		}
+	}
+}
+
+func (h *endpointHandlerImpl) Devices() map[string][]pluginapi.Device {
+	devs := make(map[string][]pluginapi.Device)
+
+	h.store.Range(func(k string, s *synchronizedEndpoint) {
+		devs[k] = s.e.Store().Devices()
+	})
+
+	return devs
+}
+
+func (h *endpointHandlerImpl) Endpoint(resourceName string) (endpoint, bool) {
+	return h.store.Endpoint(resourceName)
+}
+
+func (h *endpointHandlerImpl) Stop() error {
+	syncChan := make(chan string)
+
+	var endpoints []*synchronizedEndpoint
+	h.store.Range(func(k string, s *synchronizedEndpoint) {
+		endpoints = append(endpoints, s)
+	})
+
+	for _, s := range endpoints {
+		s.done = syncChan
+		s.e.Stop()
+
+		select {
+		case <-syncChan:
+			continue
+		case <-time.After(time.Second):
+			return fmt.Errorf("Could not stop endpoint %s", s.e.ResourceName())
+		}
+	}
+
+	return nil
+}
+
+func (h *endpointHandlerImpl) Store() endpointStore {
+	return h.store
+}
+
+func (h *endpointHandlerImpl) SetStore(s endpointStore) {
+	h.store = s
+}
+
+func (h *endpointHandlerImpl) Callback() monitorCallback {
+	return h.callback
+}
+
+func (h *endpointHandlerImpl) SetCallback(f monitorCallback) {
+	h.callback = f
+}
```

Comment:
They are here because they are needed to access the internals of the objects during tests,

Reply:
To elaborate more they are needed in tests because they allow to stub the underlying internals.
So more of a good testing practice rather that over engineering.

Looking a bit more `Callback()` is never called so it can be removed.
`Range()` is a bit similar as it allows for simplification internally and for testing

Reply:
I see.